google/cloud/spanner_v1/pool.py

```diff
@@ -138,6 +157,115 @@ def session(self, **kwargs):
         """
         return SessionCheckout(self, **kwargs)
 
+    def startCleaningLongRunningSessions(self):
+        """Starts background task for recycling session."""
+        from google.cloud.spanner_v1._helpers import (
+            DELETE_LONG_RUNNING_TRANSACTION_FREQUENCY_SEC,
+            DELETE_LONG_RUNNING_TRANSACTION_THRESHOLD_SEC,
+        )
+
+        if (
+            not AbstractSessionPool._cleanup_task_ongoing_event.is_set()
+            and not AbstractSessionPool._cleanup_task_ongoing
+        ):
+            AbstractSessionPool._cleanup_task_ongoing_event.set()
+            if self._database.logging_enabled:
+                self._database.logger.info("95% of the session pool is exhausted")
+            background = threading.Thread(
+                target=self.deleteLongRunningTransactions,
+                args=(
+                    DELETE_LONG_RUNNING_TRANSACTION_FREQUENCY_SEC,
+                    DELETE_LONG_RUNNING_TRANSACTION_THRESHOLD_SEC,
+                ),
+                daemon=True,
+                name="recycle-sessions",
+            )
+            background.start()
+        else:
+            AbstractSessionPool._cleanup_task_ongoing_event.set()
+
+    def stopCleaningLongRunningSessions(self):
+        """Stops background task for recycling session."""
+        AbstractSessionPool._cleanup_task_ongoing_event.clear()
+
+    def deleteLongRunningTransactions(
+        self, longRunningTransactionFrequency_sec, longRunningTransactionThreshold_sec
+    ):
+        """Recycles sessions with long-running transactions
+        :param longRunningTransactionFrequency_sec: Interval for running background task in seconds.
+        :param longRunningTransactionThreshold_sec: Timeout for recycling sessions in seconds.
+        """
+        long_running_transaction_timer = time.time()
+        transactions_closed = 0
+        while AbstractSessionPool._cleanup_task_ongoing_event.is_set():
+            AbstractSessionPool._cleanup_task_ongoing = True
+            is_timeout_reached = (
+                time.time() - long_running_transaction_timer
+                >= longRunningTransactionThreshold_sec
+                + longRunningTransactionFrequency_sec
+            )
+
+            if is_timeout_reached and transactions_closed == 0:
+                break
+            iteration_start = time.time()
+
+            # Retrieve a list of sessions to delete.
+            sessions_to_delete = [
+                session
+                for session in self._borrowed_sessions
+                if (datetime.datetime.utcnow() - session.checkout_time)
+                > datetime.timedelta(seconds=longRunningTransactionThreshold_sec)
+                and not session.long_running
```

Comment:
1. Should we re-name `session.long_running` to `session.eligible_for_long_running`? What we actually mean to say is if the session is allowed to be long running or not. If not (PDML & partitioned reads) then we clean it. While I fully understand the background behind this naming, a new reader will think `session.long_running` means that a session is long running. And then this check looks confusing that we are cleaning it when session is "not long running"
2.  Suggestion : It will be better to have `not session.long_running` as the first check so that we short circuit for a few times (PDML/partitioned reads)